WordPress/gutenberg
# Update theme.json shape: make settings & style top-level keys

Part of https://github.com/WordPress/gutenberg/issues/28163

This PR reorganizes the structure of theme.json to make `settings` and `styles` top-level keys, while the contexts are sub-properties of them.

Before this change: 

```
{
  "global": {
    "settings": { ... },
    "styles": { ... }
  },
  "core/paragraph": {
    "settings": { ... },
    "styles": { ... }
  }
}
```

After this change:


```
{
  "settings": {
    "global": { ... },
    "core/paragraph": { ... }
  },
  "styles": {
    "global": { ... },
    "core/paragraph": { ... }
  }
}
```

The rationale for this change is that the use cases for theme.json have grown beyond initially considered, and the vision is now being able to absorb a lot of things that themes declare at the moment via other means. Some examples are [registering (and translate?) custom templates](https://github.com/WordPress/gutenberg/pull/27778), declare theme metadata that is currently stored in the stylesheet, declare stylesheet paths, etc.

As things are ported/added to theme.json, a potential future theme.json could be like this:

```
{
  "name": "TwentyTwentyOne",
  "description": "...",
  "customTemplates": ...,
  "textDomain": ...,
  "version": 1,
  "settings": {
    "global": { ... },
    "core/paragraph": { ... }
  },
  "styles": {
    "global": { ... },
    "core/paragraph": { ... }
  }
}
```


Tasks:

- [x] Update core's experimental-theme.json
- [x] Migrate WP_Theme_JSON.
- [x] Migrate WP_Theme_JSON_Resolver
- [x] Migrate global-styles.php
- [x] Update site editor logic (at the client).
- [x] Adapt i18-theme.json to this shape https://github.com/WordPress/gutenberg/pull/28110/files#r563171543

### How to test

Test instructions:

- Use this branch from TT1-blocks for testing https://github.com/WordPress/theme-experiments/pull/175
- Make sure the front-end of the site looks right.
- Go to the post editor and fiddle with it a bit: especially test that presets, custom CSS variables, and editor configuration work as expected.
- Go to the site editor and update some user configurations (background color at the global level, etc). Make sure the changes are reflected in the editor. Save user changes and go to the front-end: you should see the same changes.
- Change some settings & styles in the theme.json of the https://github.com/WordPress/theme-experiments/pull/175 and make sure they work as expected (both front-end and editors).
